### ISSUE TITLE
add DocumentWriter.ExposeExceptions

### ISSUE THREAD

**Issue by lanwin** [graphql-dotnet/graphql-dotnet#279]
While debbuging and in test version its often very helpfull to see more infos when an error happened. Currently the DocumentWriter only exposes a simple error message like "Error trying to resolve foo". 

So I would suggest to add an option to the DocumentWriter which when set, exposes more internal information into the resolt, so its possible to guess what went wrong. This prop should be disble by default.

**Comment by jquense** [comment]
I think this is a good idea generally. `graphql-express` has `formatError` which lets the user customize the mapping of GraphQL errors to the serialized ones

**Comment by joemcbride** [comment]
Just as a note to anyone that may look at this, the errors are currently getting transformed via [ExecutionResultJsonConverter](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Execution/ExecutionResultJsonConverter.cs).

**Comment by mahald** [comment]
If u really want to expose ur exceptions u can do this like this:

```
            #if DEBUG
            List<String> Errors = new List<string>();
            foreach (ExecutionError err in result.Errors)
            {
                Errors.Add(err.ToString());
            }
            Errors.ForEach(err => {
                ExecutionError exErr = new ExecutionError(err);
                result.Errors.Add(exErr);
            });
            #endif
```

Execute it after `DocumentExecuter().ExecuteAsync` and it will expose all exeptions on Debug Mode.

**Comment by mahald** [comment]
I further improved it. I also got the wish as @jquense mentioned.
following solution let's u see fully stack trace in DEBUG mode AND Let's u throw error Messages in the Resolver that can be read:

firtst is passing any "GraphQL.ExecutionError" exception.
So u can `throw new ExecutionError("DISPLAY THIS ERROR MESSAGE");` in the resolver
second is the above and displays the full stack trace.
I replace the Errors of the result, so no double errors are in there.

```
            var result = await new DocumentExecuter().ExecuteAsync(_ => ....

            ExecutionErrors myExecutionErrors = new ExecutionErrors();

            result.Errors?.ToList().ForEach(err =>
            {
                var inner = err.InnerException?.InnerException;
                var type = inner?.GetType();
                if (type?.FullName == "GraphQL.ExecutionError")
                {
                    myExecutionErrors.Add(new ExecutionError(inner.Message));
                }
            });

            #if DEBUG
            result.Errors?.ToList().ForEach(err =>
                {
                    myExecutionErrors.Add(new ExecutionError(err.ToString()));
                });
            #endif

            if (myExecutionErrors.Count > 0)
            {
                result.Errors = myExecutionErrors;
            }

            return result;
```

so if i throw a GraphQL.ExecutionError inside the resolver now it looks like this:

```
{
  "data": {
    "permissionGroupUpdate": null
  },
  "errors": [
    {
      "message": "DISPLAY THIS ERROR MESSAGE"
    },
    {
      "message": "GraphQL.ExecutionError: Error trying to resolve permissionGroupUpdate. ---> System.AggregateException: One or more errors occurred. (DISPLAY THIS ERROR MESSAGE) ---> GraphQL.ExecutionError: DISPLAY THIS ERROR MESSAGE\r\n   at Csf4.Api.Private.ApiMutations.<>c.<<-ctor>b__0_10>d.MoveNext() in D:\\mh\\officeSVN\\csfWebBackend\\Api\\Private\\ApiMutations.cs:line 413\r\n   --- End of inner exception stack trace ---\r\n   at GraphQL.Instrumentation.MiddlewareResolver.<Resolve>d__2.MoveNext() in D:\\mh\\tmp\\graphql-dotnet-master\\src\\GraphQL\\Instrumentation\\MiddlewareResolver.cs:line 27\r\n   --- End of inner exception stack trace ---"
    }
  ]
}
```

hth

**Comment by mahald** [comment]
Well there are 2 ways of donig this:

1. change in the ExecutionResultJsonConvert for Debug Mode:
```
                #if DEBUG
                    serializer.Serialize(writer, error.ToString());
                #else
                    serializer.Serialize(writer, error.Message);
                #endif
```
The Problem here is to bring in config or parmas from somewhere

2. Add ExposeExceptions to the options of DocumentExecuter and change in DocumentExecuter.cs after Execution:

```
                    if (context.Errors.Any())
                    {
                        if (config.ExposeExceptions)
                        {
                            if (context.Errors != null)
                            {
                                result.Errors = new ExecutionErrors();
                                foreach (ExecutionError err in context.Errors)
                                {
                                    result.Errors.Add(new ExecutionError(err.ToString()));
                                }
                            }
                        }
                        else
                        {
                            result.Errors = context.Errors;
                        }
                    }
```

would u accept a PR for 2) or what do u think ?

**Comment by mahald** [comment]
Pull request #391 added to implement variant 2) as 1) doesn't work if u don't got it local

**Comment by joemcbride** [comment]
This is released in [0.17.3](https://github.com/graphql-dotnet/graphql-dotnet/releases/tag/v0.17.3)
